### PR TITLE
Skin management

### LINKED ISSUE #330: Improve the skin/cape selection UI.
### Role

I manage my skins and capes using Prism Launcher.

### Suggestion

I want the skin upload and cape management UI to be more intuitive.

### Benefit

This would provide a better experience when managing skins and capes using Prism Launcher

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

Prism Launcher should show a preview of the skin, and cape selected, when editing an account's skin/cape data using Prism Launcher. This feature occurs in the official launcher, but is absent from Prism. Currently, when selecting a cape, only the cape's name is shown, and no such preview. There is also no such preview with skins. This requires the user to go into the game just to see their changes. I feel that Prism Launcher should overhaul the skin/cape selection UI to better show the changes the user is making.

#1534 & #311: The launcher should also allow multiple skins to be saved

### LINKED ISSUE #2409: Change skin upload wording to Wide/Slim
### Role

I play Minecraft

### Suggestion

I want the skin option to say Wide/Slim instead of Steve/Alex

### Benefit

So I won't be confused

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

I didn't see any other issue made about this, so sorry if it's a duplicate. 

Anyway, while I do like how it is atm I feel like it'd create confusion for anyone doing a skin change. Took me a minute to figure out which is which, but currently minecraft shows it as wide/slim. So I feel changing the wording to match would be best

### PR DESCRIPTION
**Pull request by Trial97**
fixes #330
fixes #2409
Implemented:
- manage skins similar to icons
- visualize the 2D textures
- on the first load try to get the  data from the official Minecraft launcher
- if Prism does not manage the current skin pull it as a new skin
- saves the cape and the skin model with a specific 
- add skins from the local file
- importing from a specific user skin or from a URL

Refactored and already there:
- change skins 
- change cape
- reset skin

Not implemented and not gonna implement them in this PR:
- 3d view of texture or any fancy texture display(I have no idea on how to do that and from what I saw that would also require extra dependencies, e.g. opengl as minimum and maybe core3d for ease)
- ability to select a skin server

### CODE REVIEW

**Comment by TheKodeToad** [comment]
Looks like there's some less ambitious stuff that's not done. Should this be a draft or can those be done later?

**Comment by Trial97** [comment]
Regarding the not implemented stuff, I do not feel like doing them at all(partially because of the fact that I do not have any idea on how). Obviously I'm ok if there is somebody willing to extend this PR by implementing the mentioned stuff. As this PR solves the basic issue we can have separate PRs for the stuff I did not implemented.
Regarding the failing build I will have a look tomorrow (I'm a little tired).

**Comment by TheKodeToad** [comment]
I think using an existing player's skin should be easy

**Comment by TheKodeToad** [comment]
I don't think getting data from the official launcher is a great idea:
- it doesn't happen with anything else
- it's a proprietary format prone to change
- it won't work in sandboxed environments

**Comment by TheKodeToad** [comment]
Removing that would also fix actions (although there's probably something that works with Qt5)

**Comment by Trial97** [comment]
Ok removed the default launcher stuff.
Added more import options(from a username and from a URL)
Some open items: 
- Do I need to check if two skins are the same? If yes any fast way of doing that?
- I always rewrite the files on import; Do I need to implement some sort of protection for that or can I leave it as is?
- I know that I'm not good at designing GUI so do not expect this to look so good(I need a little help on that)

**Comment by TayouVR** [comment]
> 3d view of texture

I wanted to implement this, with support for ears x3
Maybe if I have the time and motivation for it i could look into it after this PR.

**Review by TayouVR — APPROVED** [review]
I tried renaming a skin and it duplicated it, keeping the original file.
can't reproduce it though and log doesn't say anything about it either as far as i can see.

Also just looked through the code to see if I can find how that could happen... 
does FS::move perform a copy if the original file can't be deleted for whichever reason?

Edit: I can reproduce it. If the Skin has a url attached it seems to duplicate when renaming

image
owo and Ears_Wings2 are both from renaming the one with the id as name

Edit2: I am also not able to delete that one in the UI;
prismlauncher_imSSiLpQvk

Edit3: also renaming causes the playermodel (and cape presumably) to be reset, I assume because the rename function just renames the file, it doesn't actually take care of updating the json?

**Review by TayouVR — CHANGES_REQUESTED** [review]
renaming needs to be handled better, whatever is up with the user-id based skin not being possible to be deleted needs to be changed or better communicated.

**Comment by Trial97** [comment]
Yeah, you can not remove the current skin. Let me see what I can do regarding that and the rename stuff.

**Comment by Trial97** [comment]
@TayouVR I should have addressed all the mentioned issues:
- for delete: if it is the selected skin a warning message will pop up
- the rename was fixed (even for current selected skin)

**Comment by TayouVR** [comment]
- deleting currently selected, warning works 👍 
- renaming it, works 👍 

but it seems whenever I open the window the model is set to classic even though its saved as slim
notepad++_21wOospjDR

**Comment by Trial97** [comment]
Fixed that too. It was a case mismatch the API returned "SLIM" I checked for "slim".

**Review by TayouVR — APPROVED** [review]

File: launcher/ui/dialogs/skins/SkinManageDialog.ui [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+          </sizepolicy>
+         </property>
+         <property name="title">
+          <string>Model</string>
+         </property>
+         <layout class="QHBoxLayout" name="horizontalLayout_2">
+          <item>
+           <widget class="QRadioButton" name="steveBtn">
+            <property name="text">
+             <string>Clasic</string>
```
Comment (DioEgizio) [inline launcher/ui/dialogs/skins/SkinManageDialog.ui]:
```suggestion
             <string>Classic</string>
```

**Comment by Trial97** [comment]
@getchoo  you already said you will review 
image

**Review by getchoo — CHANGES_REQUESTED** [review]
most of this LGTM and works well. just two notes:

- errors are silently ignored. they probably shouldn't be
  - this is very confusing when trying to import the skin of another player for example. if i type the name in wrong and press "import user", i only see a flash of a loading screen that appears to do nothing
  - this also can fool you into thinking your changes applied, when in reality they didn't
- non-reproducible crash
  - when changing from the slim to classic model, the entire launcher crashed. i had sadly not launched it from the cli so no errors. i haven't been able to replicate it since, and i'd imagine this is an issue in NetJob/Request instead of being introduced here. not going to make this blocker due to that, but something i figured i should bring up

**Comment by Trial97** [comment]
Added the error message for that buttons, but could not produce any crash so nothing on that front

**Review by getchoo — CHANGES_REQUESTED** [review]
everything else seems good. last issue i've found is that the error messages do display when importing a user's skin, but not when importing from a url

File: launcher/ui/dialogs/skins/SkinManageDialog.cpp [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+            qCritical() << "Couldn't load skin json:" << e.cause();
+        }
+        uuidLoop->quit();
+    });
+
+    connect(getProfile.get(), &Task::succeeded, this, [profileLoop, profileOut, job, getProfile, &mcProfile, downloadSkin] {
+        if (Parsers::parseMinecraftProfileMojang(*profileOut, mcProfile)) {
+            downloadSkin->setUrl(mcProfile.skin.url);
+        } else {
+            job->abort();
```
Comment (Ryex) [inline launcher/ui/dialogs/skins/SkinManageDialog.cpp:469]: should this branch report the error to the user?

**Review by Ryex — CHANGES_REQUESTED** [review]

**Comment by Ryex** [comment]
other than those two points things look good code wise and I think the interface looks good too

File: launcher/ui/dialogs/skins/SkinManageDialog.cpp (hunk 2 above)
Comment (Trial97) [reply launcher/ui/dialogs/skins/SkinManageDialog.cpp:469]: No, as I consider all steps a whole task here: I display a generic fail message at the end.

**Comment by Trial97** [comment]
> other than those two points things look good code wise and I think the interface looks good too

you pointed out just one. What is the other one

File: launcher/ui/dialogs/skins/SkinManageDialog.cpp [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+    auto profileLoop = makeShared<WaitTask>();
+
+    auto getUUID = Net::Download::makeByteArray("https://api.mojang.com/users/profiles/minecraft/" + user, uuidOut);
+    auto getProfile = Net::Download::makeByteArray(QUrl(), profileOut);
+    auto downloadSkin = Net::Download::makeFile(QUrl(), path);
+
+    connect(getUUID.get(), &Task::aborted, uuidLoop.get(), &WaitTask::quit);
+    connect(getUUID.get(), &Task::failed, uuidLoop.get(), &WaitTask::quit);
+    connect(getProfile.get(), &Task::aborted, profileLoop.get(), &WaitTask::quit);
+    connect(getProfile.get(), &Task::failed, profileLoop.get(), &WaitTask::quit);
```
Comment (Ryex) [inline launcher/ui/dialogs/skins/SkinManageDialog.cpp]: I think the `failed` events should use some logging and or inform the user of the failed task

File: launcher/ui/dialogs/skins/SkinManageDialog.cpp (hunk 2 above)
Comment (Ryex) [reply launcher/ui/dialogs/skins/SkinManageDialog.cpp:469]: I take it you mean the error if the resulting skin after all the tasks end isn't valid?

Comment (Trial97) [reply launcher/ui/dialogs/skins/SkinManageDialog.cpp:469]: yes: if the task fails in any way the skin is not downloaded to path making it invalid

**Review by Ryex — APPROVED** [review]

**Review by getchoo — APPROVED** [review]
awesome work

**Comment by bittin** [comment]
and now translated to Swedish :)

**Comment by bedxnta** [comment]
would appreciate a few more pics to see how it looks :3

**Pull request merged**